backend/catalogue/management/commands/generate_layer_previews.py

```diff
@@ -0,0 +1,119 @@
+from django.core.management.base import BaseCommand
+from django.core.files.storage import default_storage
+from django.core.files import File
+from django.conf import settings
+from urllib.request import urlopen
+from urllib.parse import urlencode
+from PIL import Image
+from io import BytesIO
+
+from catalogue.models import Layer
+
+
+basemap = Image.open(default_storage.open('basemap.png'))
+
+
+def basemap_latitude_to_pixel(latitude):
+    t = 0.5 + (latitude / 360)
+    return round(basemap.width * t)
+
+
+def basemap_longitude_to_pixel(longitude):
+    t = 0.5 - (longitude / 180)
+    return round(basemap.height * t)
+
+
+def basemap_bbox(north, south, east, west):
+    left = basemap_latitude_to_pixel(west)
+    right = basemap_latitude_to_pixel(east)
+    upper = basemap_longitude_to_pixel(north)
+    lower = basemap_longitude_to_pixel(south)
+
+    if (left > right):
+        left_img = basemap.crop((left, upper, basemap.width + right, lower))
+        right_img = basemap.crop((0, upper, right, lower))
+        left_img.paste(right_img, (left_img.width - right, 0))
+        return left_img
+    else:
+        return basemap.crop((left, upper, right, lower))
+
+
+def retrieve_image(layer):
+    bbox = {
+        'north': float(layer.maxy),
+        'south': float(layer.miny),
+        'east':  float(layer.maxx),
+        'west': float(layer.minx)
+    }
+
+    x_delta = bbox['east'] - bbox['west']
+    y_delta = bbox['north'] - bbox['south']
+    aspect_ratio = x_delta / y_delta
+    width = 386
+    height = round(width / aspect_ratio)
+
+    params = {
+        'service': 'WMS',
+        'request': 'GetMap',
+        'layers': layer.layer_name,
+        'styles': '',
+        'format': 'image/png',
+        'transparent': 'true',
+        'version': '1.1.1',
+        'width': width,
+        'height': height,
+        'srs': 'EPSG:4326',
+        'bbox': '{west},{south},{east},{north}'.format(**bbox)
+    }
+
+    url = '{server_url}?{params}'.format(
+        server_url=layer.server_url, params=urlencode(params))
+
+    layer_image = Image.open(urlopen(url))
+    cropped_basemap = basemap_bbox(**bbox).resize((width, height))
+    cropped_basemap.paste(layer_image, None, layer_image)
+    return cropped_basemap
+
+
+class Command(BaseCommand):
+    def add_arguments(self, parser):
+        parser.add_argument(
+            '--layer_id'
+        )
+
+    def handle(self, *args, **options):
+        layer_id = int(options['layer_id']
+                       ) if options['layer_id'] != None else None
+
+        if layer_id is not None:
+            layer = Layer.objects.get(id=layer_id)
+
+            bytes_io = BytesIO()
```

Comment:
Minor point of style -- have you see the python "with" statement?  (in this scenario it would basically be a try-finally; automatically closing the resource)